### DISCUSSION TITLE
Achievement Variables / Memory

### DISCUSSION THREAD

**Discussion by gdeOo** [RetroAchievements/RAIntegration#780] [Ideas]
This is a toolkit feature request / suggestion / discussion.

The aim is to allow achievements to have their own memory and versatile ways of manipulating it. This would allow for triggers based on past events that the game itself may not "remember".

This is a long post, so here's a sneak preview:
- first a proposal based on hit targets;
- then a better (IMO) proposal based on variables as a new concept;
- finally a proposal for value manipulation operators.

----

### Option 1: Hit targets as variables

Achievement developers can already work with their own memory using Hit Targets, which can be thought of as variables in a programming sense.

These variables are, however, very limited in how they can be manipulated: they can only be incremented by 1 or reset to 0. They can also be incremented by more than 1 using AddHits, though with additional caveats.

#### SubHits

The simplest extension would be the creation of a `SubHits` flag. Just by itself this flag would enable much more complex use of hit targets as variables.

Motivating example that I ran into myself:
  * Scenario:
    * Game has a puzzle in which the player can operate 6 consoles, 3 of which are on the left and the other 3 are on the right.
    * The achievement reads like this: "Solve the puzzle without operating 2 consoles on the same side twice in a row".
    * The game does not remember which console was operated last in its own memory.
  * Current solution: a very obtuse abuse of the toolkit based on a bunch of alts, hit targets, resets and pause ifs.
  * Solution with `SubHits`: just 2 alts, one for each side. Alt for the right side would be:
```
AddHits 0x<operated right console> = 1
SubHits 0x<operated left console> = 1
SubHits 1 = 1                   // ugly hack for the fact that
ResetIf 1 = 1 (Hit target = 2)  // the non-hits cond must be true
```

#### (Add|Sub)SourceAsHits

Add the currently accumulated value (from a simple memory access, from a sequence of AddSources, etc...) to the Hit count.

This would enable additions to and subtractions from the variables (hit targets) without having to rely on a fixed-length chain of (Add|Sub)Hits.

----

### Option 2: Variables as a new concept

Even though we can "abuse" hit targets for our variables, I think it might be preferable to introduce them as a new concept.

I think the following approach could work with the current achievement serialization & editor:
  * Each achievement could be allowed the use of N integer variables, accessible by an index, as though they were in a vector.
  * All variables initialized to zero upon achievement reset.
  * New "access type" `Var` to use the value of a variable in other conditions, i.e., `Var` instead of `Mem`, `Delta`, `Value`, ...
```
         Var(0) = 123  // triger if variable 0 = 123
ResetIf  Var(0) < 10   // reset if variable 0 < 10
```
  * `StoreVar` flag to store the currently accumulated value into the variable with a given index. Left side of the condition would provide a value or a memory reference, whilst the right side would provide the index.
```
StoreVar   Mem(0xabcd)  Var(0)  // variable 0 = value at 0xabcd

StoreVar   Value(1234)  Var(0)  // variable 0 = 1234

AddSource  Mem(0xabcd)
StoreVar   Value(1234)  Var(0)  // variable 0 = (value at 0xabcd) + 1234

AndNext   Mem(0xabcd) != Value(0)  // if value at 0xabcd is not zero...
StoreVar  Mem(0xabcd)    Var(0)    //   store it in variable 0
```
  * `AddToVar` & `SubFromVar` with obvious semantics similar to `StoreVar`.

#### Bonus: variables as the main avenue for multi session

The set of variables used by an achievement could be the perfect thing to store in order to support multi session achievements.

Since variables are named (through their index), they have a consistent meaning across revisions of an achievement, which is not the case for hit targets. Developers can leverage this fact to ensure changes to a multi session achievement are backwards compatible, for example.

----

### Bonus: More value manipulation operators

If variables become a thing, the only thing keeping the toolkit from nearly becoming a full blown programming language is the lack of ways to manipulate values. I've seen some other issues that request specific value manipulation instructions: #417, #184, maybe more.

To keep the achievement editor as is, operations on values would probably have to be implemented as conditions with operation-specific Flags operating on the currently accumulated value. For binary operations, the right side would provide the second operand. The resulting value would then become the currently accumulated value, available for the next condition to use.

```
Multiply  Mem(0xabcd)  Val(2)  // inlined 1st and 2nd operands

AddSource Value(1)
Multiply  Mem(0xabcd)  Val(2)  // accumulated + inlined

ToBool    Mem(0xabcd)          // unary operator
```

There might be a problem here in that I'm not sure if the existing conditions can have empty left sides. If they can't, we could still use the value manipulator operations like an AddSource though. For example:

```
ToBool   Mem(0xabcd)
ResetIf               = 1  // might not be ok

ToBool   Mem(0xabcd)
ResetIf  Value(0)     = 1  // awkward, but ok because ToBool acts like an AddSource
```

Some useful operator off the top of my head: `Multiply`, `Divide`, `Remainder`, `Abs`, `ToBool`, `Shift(Left|Right)`, `BinaryAnd`, `BinaryOr`, `BinaryXor`, `BinaryFlip`.

----

I'm sure I haven't thought of everything. If there's interest, let's iterate on this.

Also, I only provided one motivating example, but I can provide more if necessary.

**Comment by televandalist** [comment]
Multi-session is, without a doubt, my most wanted feature ~~that isn't screen rotation~~. Not only for achievements involving long-term goals, but I feel like "a password used must be a password earned" could be made possible with it. 

<3 <3 <3

**Comment by Keltron3030** [comment]
I used to express the need for variables early on, but with the coming of andnext the need has lessened somewhat in my opinion.  There was previously talk of incorporating Lua scripting but some felt that would be overkill for our needs.

There is of course always the consideration of ensuring compatibility with Retroarch and any solution of this kind must ensure that compatibility is maintained given a large swath of our users rely on that software to play on their various devices, myself included.  

In any case I am always in favor of a solution that expands what's possible, like adding multi-session for example.  My only desire would be for the more advanced features to be implemented with an intuitive interface since the biggest benefit to the current tools is they are relatively accessible to people of all skill levels and makes for a low access barrier where "anyone can make achievements".

**Comment by Jamiras** [comment]
I've imagined variables in a slightly different way. Each variable would behave like a leaderboard value. It could be a constant (to handle regional differences), mathematically computed or be a resettable HitCount counter. Variables would be updated once each frame (before evaluating the achievements/leaderboards which might reference them). Each variable could be referenced by multiple achievements to avoid duplication.

And as you've suggested, variables could be marked to be stored between sessions.

My thoughts on #417 were that `Multiply` would not be a flag. Values that can be multiplied would have no left side, and the "blank" right side would be replaced with the multiplication logic. You might do something like this to check if a value multiplied by 5 is greater than 100 (though it would be simpler to just compare the value to 20):
```
AddSource Mem(0xabcd) * 5.0
          0 > 100
```
The multiplier would be a floating point value to simulate division and for consistency with leaderboard values.

Regarding the additional operators:
* `SubHits` seems reasonable enough and is a good complement to `AddHits`
* `AddSourceAsHits`/`SubSourceAsHits` - I'd need to see a use case for this where it's desirable to modify a HitCount by a memory value. Variables would be a better way to capture a complex value. Then you'd compare the variable to a constant instead of comparing a HitCount to a target HitCount.
* `Multiply` and `Divide` would be handled through `AddSource`, `SubSource`, `AddHits` and `SubHits` via #417. `ShiftLeft` and `ShiftRight` are just multiply/divide operations with power of 2 operands. I'd expect users to just use the `Multiply` and `Divide` functions, as most developers aren't programmers and probably wouldn't understand bit shifting. See also my later note on bit operations.
* `Remainder` and `Abs` would be nice to haves. I've already implemented `Abs` once using existing semantics. I'd recommend renaming them to `ModulusSource` and `AbsSubSource`.
* The `Binary` operators would have minimal use as we already allow direct access to individual bits. Any logic that could be applied across multiple bits via the Binary operators could be applied to the individual bits and ANDed or ORed together using multiple conditions/alt groups/`AndNext`/`OrNext`. Most of the proposed operators modify the data, so would only be applicable to generating variables. When processing triggers, you just look at memory.

**Comment by gdeOo** [comment]
I was mostly interested in variables as a way of memorizing prior data/events that the game does not itself memorize. If I understood correctly, the way you've imagined them, @Jamiras, has more to do with reusing expressions, though I'm kinda suspicious that I did not fully grasp your idea.

The HitCount counter you mentioned could be used as a variable in the sense I suggested, but only if `(Add|Sub)SourceAsHits` were implemented. Otherwise we're still limited to the unitary increments + reset to 0 logic. A variable that is not bound to this limitation would be preferable IMO.

----

Regarding the scenario for using a memory value as hits, here's an example I actually ran into:
  * Puzzle in which the player is allowed 1 to N moves per turn
  * Achievement I want: "Solve the puzzle in 6 moves or less"
  * Game has a byte that contains the number of moves used in the current turn, but does not remember how many moves were used in total

Using actual variables, I was imagining this solution:
```
ResetIf   0x<in_puzzle>              = Value(0)   // only process when in the puzzle (also reset variables to 0)
AndNext   Mem(0x<turn_just_started>) = Value(1)   // if turn just started:
AddSource Mem(0x<current_num_moves>)
StoreVar  Var(0)                       Var(0)     //   total_moves = total_moves + current_num_moves
ResetIf   Var(0)                     > 6          // reset if total_moves > 6
          Mem(0x<puzzle_solved>)     = Value(1)   // trigger if puzzle is solved
```

The HitCount based solution would be something like this:
```
ResetIf         0x<in_puzzle>              = Value(0)
AndNext         Mem(0x<turn_just_started>) = Value(1)
AddSourceAsHits Mem(0x<current_num_moves>)
SubHits         Value(1)                   = Value(1)
ResetIf         Value(1)                   = Value(1)  (Hit Target = 7)
                Mem(0x<puzzle_solved>)     = Value(1)
```

Due to the small numbers involved, this scenario can be solved with the current toolkit, but I think it still illustrates my intent well.

How would this solution be with the way you imagined variables?

----

Regarding the value manipulation operators, they were a bit of an after-thought, and I'm fine with everything you said @Jamiras. The only advantage to having them all would be clearer intent when reading an achievement's code, but the number of flags could become unnecessarily overwhelming.

----

I had not considered sharing variables across achievements, though it makes perfect sense (for the password scenario that @Televandalist mentioned, for example). But I also see many variables being useful in a single achievement, so maybe we could have local & global variables?

Global variables would have to be explicitly reset, instead of reseting along with the cheevo. There would also be an issue with inter-achievement sincronization if `StoreVar` were allowed on them. Maybe @Jamiras' idea of calculating variables every frame before evaluating achievements would work better in this case.

----

Regarding the interface, @Keltron3030, I'm sure there would be more intuitive ways of implementing this. My suggestion was meant to work inside the limits of the current achievement editor, because I get the impression that changing it is not a good idea. It would also have the least impact on developers that wouldn't want to use these features.

@Jamiras' idea of having variables separate from achievements would probably need a separate interface though, so maybe there's an opening to do things outside the constraints of the editor.

**Comment by lonegraywolf2000** [comment]
I came across a situation where this would have been useful.

In Dance Dance Revolution 4th Mix, there are two variables (per player, focusing on one) for the running score: the only difference is that one is instant and the other is used for a rolling meter of sorts with the UI. The problem is that this variable doesn't reset between stages like I thought it did. It was instead incrementing between stages, and yet the game UI still displayed the "intended" correct value. The intended value was never found in the expected parts of memory. As far as how Konami did this, I don't know yet.

A situation like this would be perfect for achievement variables. I could note the score at the Start condition, note at the Submit condition, and do math on the Value condition.

At least, that's what I'd like to be able to do. I currently have a hacked leaderboard setup that should work for the time being, but is technically brittle.

**Reply by redwizard42** [reply]
Sounds also like the idea of AddSourceAsHits would work for this.  You can measure off hits in the value, and you'd so something like...

SubSource Delta [ScoreAddress]
AddSourceAsHits Mem [ScoreAddress]
Measured Value 1 = Value 0

And so value would be the total of every score increase during the leaderboard.

I wanted something like this for Individual Level scores for a Gradius game I did, where I wanted the value to be Current Score-ScoreWhenEnteringLevel

**Comment by suXinjke** [comment]
I want to share ideas regarding multisession, I guess this thread fits.

This old Variables proposal - I don't think it's viable because I see risks of losing progress by editing the set. Right now, if you edit the achievement - the hits are not preserved when loading save state which makes sense to me - the logic for those hits may be considered invalid now. I think majority of people consider variables/multisession for really long sessions, therefore risk of losing progress should be close to zero.

When someone was dreaming of multisession in Discord *(again?)* I got an idea for entirely different approach which I see may fit well into existing RAWeb database schema - so there's no need for custom variable format or new code that set developers would have to learn.

## Proposal

**Let the client report triggers for lesser achievements that contribute towards the actual achievement.** In other words, allow to define parent achievement that can have sub/children achievements in it. You have to trigger all sub-achievements to have parent achievement triggered.

The way I see it - it would only require changing achievement table to allow linking them to the parent, changes to backend so it triggers the parent achievement at proper time, changes to frontend. What's interesting is that technically - even old rcheevos and RAIntegration should kinda work with these because they will perceive them as regular 0 pointer achievements anyway (since they come from the same table with same structure, while ignoring the parent achievement id). Without user updating the libraries, they will end up less user friendly with unlock spam and making the original list look confusing, but users will eventually update.

Let's talk examples: "Complete all 4 Levels on hard difficulty" kind of achievements. If game tracks the progress, classic implementation can involve trusting the game flags in-memory, usually with AddSource. So when you go from `< 4` to `== 4`, you trigger the achievement. This can be problematic because save file can be tampered so you don't have to do the whole thing. You can prevent that if you make achievement for each Level instead, making it 4 achievements instead of 1 and less save tampering opportunities.

Problem arises when you have too many levels or whatever, like 25 of them or hundreds, such amount would bloat the set. I see people usually resort to grouping them while having faith in honor system.

This parent/sub achievement system could help solve this problem, you would code achievement for each level, but user perceives them as progress towards the main achievement, the website and emulators should display it as so, kinda like this:

![image](https://user-images.githubusercontent.com/9765719/201297323-0f374ce9-68c6-44f5-897d-2d867b4a60f0.png)

### Caveats

The caveats would be mostly regarding editing such parent/sub groups. Imagine a scenario of parent requiring you to do 4 sub achievements, and I got 3 of them. Set author then decides that the 4th one I didn't get should not be part of the group - if this sub achievement goes out of the group and/or into unofficial - do we reward me the parent achievement now? What about adding new sub achievements into these groups? Should they deprave you of original unlock? Issue is similar to mastery badges and I don't have exact solution in mind for now.

### Complex case: sequential unlocking

Let's suppose you have to finish a big game on Hard difficulty *without doing game breaking glitches that would skip the game*. If you were to specify sub-achievements *as sequence* that must be unlocked exactly in specified order - this could help with maintaining that. I'm not sure how user friendly this would be, like if user misses portion of the sequence - will they notice? It's also questionable on what to do with the sequence progress - should it reset automatically if I trigger the first portion of sequence again, or should it reset manually only? If people have old rcheevos and emulators, they will see such achievements trigger, while backend will not mark them as unlocked (because they were unlocked in incorrect order). This would also require additional column in achievement table, unless you'd make sub-achievement to have another sub-achievement as parent which I personally find extremely confusing.

### Complex case: counting

Let's take an example of "Kill X demons with Y". Achievements may ask you to do this in one session because the game doesn't track total amount of kills, this is ok solution but it doesn't allow to specify really big amounts - player would have to keep the emulator on. Even if the game did track the total killcount, it still can be tampered with a savefile. Solution: allow the achievement trigger many times and send the amount of triggers to the backend, so backend adds the reported amount to the stored counter. This sure won't be compatible with old rcheevos, it may trigger only once and user will go *wtf*. Backend could choose to not send send these as part of the set list for compatibility purposes. If one were to implement this, the data sending should be debounced as several triggers may occur in very short time. Database wise, one could include `count` column to `awarded` table for this purpose. Regular achievements would have `count` as 1. Such achievements will make the point calculation queries more complex - they will also have to compare that count does match the required one.

Emulators can take inspiration from Steam overlay displaying the progress for such achievements in the corner. Like it'd show up every 100 kills if you're required to do 1000.

Additionally, maybe it could also accumulate deltas? If one wants to make an achievement "Amass X amount of prize money" - you wouldn't like to track the total amount of money in-game because you can tamper the save file, you can make rcheevos send money increments which still go towards regular trigger count.

**Reply by Blazekickn** [reply]
yes please

**Reply by Jamiras** [reply]
sub achievements are discussed in #768

> It has also been proposed that the sub-achievements be non-visible and serve as multi-session checkpoints, in which case the sub-achievement would be worth 0 points. There are probably better ways to implement multi-session checkpoints.

**Comment by televandalist** [comment]
![Screenshot_20221111-034345](https://user-images.githubusercontent.com/45054151/201313384-dd1e24f9-7765-4dde-9e1b-c13ca3a8b013.png)

wen

**Reply by suXinjke** [reply]
I vaguely remember some message in now deleted #pcsx2-rollout made me realise "this could work", maybe it was yours too, so idea is not purely mine, but atleast it's properly outlined here.

**Comment by televandalist** [comment]
I think I saw @Sanaki mention it at one point and I haven't stopped thinking about it ever since. 😆
